### ISSUE TITLE
feat: GPU-Resident History Module

### ISSUE THREAD

**Issue by fbanelli** [antonioterpin/goggles#5]
# [Feature Request] GPU-Resident History Module

## Summary

Introduce a **generic JAX-based history subsystem** inside **Goggles**, designed for reusable, GPU-resident temporal buffers. The module should generalize temporal memory handling currently embedded in FlowGym's `EstimatorState`, making it a standalone component usable across multiple projects.

---

## Motivation

FlowGym currently manages short-term temporal memory (e.g., image pairs, flow estimates) within its estimator states. This logic is hardcoded and not reusable outside FlowGym, despite being a generic need in GPU-based pipelines.

Moving this functionality into Goggles aligns with its role as a **GPU-first data handling and visualization framework**, enabling consistent, device-local temporal structures.

### Benefits

- **Reusability:** Common GPU memory pattern for FlowGym, RL agents, and other JAX pipelines.
- **Performance:** On-device buffer updates and slicing; avoids host synchronization.
- **Modularity:** Clean separation between algorithm logic (FlowGym) and memory management (Goggles).
- **Extensibility:** Unified design for histories, replay buffers, or feature sequences.

---

### Core Classes

```python
@dataclass
class HistoryFieldSpec:
    length: int
    shape: tuple[int, ...]
    dtype: jnp.dtype = jnp.float32
    init: str = "zeros"

@dataclass
class HistorySpec:
    fields: dict[str, HistoryFieldSpec]

    @classmethod
    def from_config(cls, config: dict[str, Any]) -> "HistorySpec":
        return cls(fields={k: HistoryFieldSpec(**v) for k, v in config.items()})
```

### Creation and Update Utilities

```python
def create_history(spec: HistorySpec, batch_size: int, rng: PRNGKey) -> dict[str, jnp.ndarray]:
    """Allocate GPU-resident history tensors following (B, T, *shape)."""

def update_history(history: dict[str, jnp.ndarray], new_data: dict[str, jnp.ndarray]) -> dict[str, jnp.ndarray]:
    """Shift and append new items along the temporal axis."""
```

### Shape Convention

Each field follows:

```
(B, T, ...)
```

- **B:** Batch dimension (parallel environments, samples, etc.)
- **T:** Temporal dimension (history length)
- Independent `T` per field supported.

---

## Integration Plan

1. **New module:** `goggles.history` with submodules:
   - `spec.py` — schema and validation
   - `buffer.py` — initialization and update logic
   - `utils.py` — optional padding, slicing, or stacking helpers
2. **FlowGym migration:** Replace local `history_*` fields in `EstimatorState` with `from goggles import history`.
3. **Backwards support:** Provide a compatibility adapter for existing FlowGym checkpoints.

## Sub-issues

1. Module scaffolding and API design  (#6 )
   - Create `goggles/history` package, add `spec.py`, `buffer.py`, `utils.py`, type hints and minimal tests.

2. Core buffer implementation (JAX)  (#7 )
   - Implement `create_history` and `update_history` using `jax.numpy` and `jax.lax` for in-device shifting, ensure jittable ops and no host-device copies.

3. Validation, tests, and benchmarks  (#8 )
   - Unit tests for shape/dtype correctness and update semantics.

4. FlowGym migration and compatibility adapter
   - Replace local `history_*` fields in `EstimatorState` with `goggles.history` usage and provide a loader to convert old checkpoints.

---

## Next Steps

- Add `to_host` and `to_device` wrappers
